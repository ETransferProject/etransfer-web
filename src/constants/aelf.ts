import { AELF_NODES } from './index';

export const AElfNodes = AELF_NODES;
