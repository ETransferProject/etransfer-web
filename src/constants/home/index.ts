import { Deposit, Withdraw, Records, InfoDashboard } from 'assets/images';

// SideMenuKey must bind BusinessType
export enum SideMenuKey {
  Deposit = 'Deposit',
  Withdraw = 'Withdraw',
  History = 'History',
  Info = 'Info',
}

export const MENU_ITEMS = [
  {
    icon: Deposit,
    key: SideMenuKey.Deposit,
    label: SideMenuKey.Deposit,
  },
  {
    icon: Withdraw,
    key: SideMenuKey.Withdraw,
    label: SideMenuKey.Withdraw,
  },
  {
    icon: Records,
    key: SideMenuKey.History,
    label: SideMenuKey.History,
  },
  {
    icon: InfoDashboard,
    key: SideMenuKey.Info,
    label: SideMenuKey.Info,
  },
];
