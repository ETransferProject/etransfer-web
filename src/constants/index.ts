export enum WalletType {
  unknown = 'unknown',
  discover = 'discover',
  portkey = 'portkey',
}

export enum NetworkType {
  MAIN = 'MAIN',
  TESTNET = 'TESTNET',
}

export const ChainNamePrefix = {
  MainChain: 'MainChain',
  SideChain: 'SideChain',
};

export const USDT_DECIMAL = 6;
export const SECONDS_60 = 60000;

export const prefixCls = 'etransfer-web';
export const AppName = 'ETransfer Web';
export const BrandName = 'ETransfer';

export * from './testnet';
