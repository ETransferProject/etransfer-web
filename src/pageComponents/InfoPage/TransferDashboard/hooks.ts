import {
  DefaultTransferDashboardFromTokenOptions,
  DefaultTransferDashboardFromChainOptions,
  DefaultTransferDashboardToTokenOptions,
  DefaultTransferDashboardToChainOptions,
} from 'constants/infoDashboard';
import { useMemo } from 'react';
import { TTransferDashboardFilterToken, TTransferDashboardFilterNetwork } from 'types/api';

export interface UseFilterParams {
  fromTokenList: TTransferDashboardFilterToken[];
  fromChainList: TTransferDashboardFilterNetwork[];
  toTokenList: TTransferDashboardFilterToken[];
  toChainList: TTransferDashboardFilterNetwork[];
}

export function useFilter({
  fromTokenList,
  fromChainList,
  toTokenList,
  toChainList,
}: UseFilterParams) {
  const fromTokenOptions = useMemo(() => {
    const list: { value: number | string; label: string }[] = [
      DefaultTransferDashboardFromTokenOptions,
    ];
    fromTokenList.forEach((item) => {
      list.push({ value: item.key, label: item.symbol });
    });
    return list;
  }, [fromTokenList]);

  const fromChainOptions = useMemo(() => {
    const list: { value: number | string; label: string }[] = [
      DefaultTransferDashboardFromChainOptions,
    ];
    fromChainList.forEach((item) => {
      list.push({ value: item.key, label: item.network });
    });
    return list;
  }, [fromChainList]);

  const toTokenOptions = useMemo(() => {
    const list: { value: number | string; label: string }[] = [
      DefaultTransferDashboardToTokenOptions,
    ];
    toTokenList.forEach((item) => {
      list.push({ value: item.key, label: item.symbol });
    });
    return list;
  }, [toTokenList]);

  const toChainOptions = useMemo(() => {
    const list: { value: number | string; label: string }[] = [
      DefaultTransferDashboardToChainOptions,
    ];
    toChainList.forEach((item) => {
      list.push({ value: item.key, label: item.network });
    });
    return list;
  }, [toChainList]);

  return {
    fromTokenOptions,
    fromChainOptions,
    toTokenOptions,
    toChainOptions,
  };
}
