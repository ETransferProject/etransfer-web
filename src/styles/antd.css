html {
  --ant-primary-color: #285eff;
  --ant-primary-color-hover: #3b6cff;
  --ant-primary-color-active: #1e52f0;
  --ant-success-color: #00b73e;
  --ant-error-color: #f53f3f;
  --ant-error-color-hover: #fc7373;
  --ant-warning-color-active: #cf3737;
  --ant-warning-color: #f90;
  --ant-info-color: #285eff;
}
