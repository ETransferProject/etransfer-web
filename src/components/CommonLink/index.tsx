import LinkOutlined from '@ant-design/icons/LinkOutlined';
import { Button } from 'antd';
import React from 'react';
import { isUrl } from '../../utils/reg';
import clsx from 'clsx';
import { useCommonState } from 'store/Provider/hooks';

export default function CommonLink({
  href,
  children,
  className,
  showIcon = true,
  isTagA,
}: {
  href: string;
  children?: React.ReactNode;
  className?: string;
  showIcon?: boolean;
  isTagA?: boolean;
}) {
  const { isMobile } = useCommonState();
  const target = isMobile ? '_self' : '_blank';
  if (isTagA)
    return (
      <a target={target} href={href} rel="noreferrer">
        {children}
      </a>
    );
  return (
    <Button
      className={clsx('common-link', className)}
      disabled={!href || !isUrl(href)}
      onClick={(e) => e.stopPropagation()}
      target={target}
      type="link"
      href={href}>
      {showIcon && <LinkOutlined />}
      {children}
    </Button>
  );
}
