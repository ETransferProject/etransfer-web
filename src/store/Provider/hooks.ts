import { useCallback, useMemo } from 'react';
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import type { RootState, AppDispatch } from './store';
import { emitLoading } from 'utils/events';
import { LoadingProps } from 'components/Loading';
import { setIsUnreadHistory } from 'store/reducers/common/slice';
import { resetDepositState } from 'store/reducers/deposit/slice';
import { resetWithdrawState } from 'store/reducers/withdraw/slice';
import { resetRecordsWithoutFilter } from 'store/reducers/records/slice';
// import { resetInfoDashboardState } from 'store/reducers/infoDashboard/slice';

// Use throughout your app instead of plain `useDispatch` and `useSelector`
export const useAppDispatch: () => AppDispatch = useDispatch;
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;

export const useCommonState = () => useAppSelector((state) => state.common);
export const useDepositState = () => useAppSelector((state) => state.deposit);
export const useWithdrawState = () => useAppSelector((state) => state.withdraw);
export const useWithdrawNewState = () => useAppSelector((state) => state.withdrawNew);
export const useCrossChainTransfer = () => useAppSelector((state) => state.crossChainTransfer);
export const useRecordsState = () => useAppSelector((state) => state.records);
export const useInfoDashboardState = () => useAppSelector((state) => state.infoDashboard);

export const useLoading = () => {
  const _setLoading = useCallback(
    (isLoading: boolean, loadingInfo?: LoadingProps) => emitLoading(isLoading, loadingInfo),
    [],
  );
  return useMemo(() => ({ setLoading: _setLoading }), [_setLoading]);
};

export const useResetStore = () => {
  const dispatch = useAppDispatch();

  return useCallback(() => {
    // dispatch(resetCommon());
    dispatch(setIsUnreadHistory(false));
    dispatch(resetDepositState());
    dispatch(resetWithdrawState());
    dispatch(resetRecordsWithoutFilter());
    // dispatch(resetInfoDashboardState());
  }, [dispatch]);
};
