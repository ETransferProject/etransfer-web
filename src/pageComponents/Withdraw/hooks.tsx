import { ZERO } from 'constants/calculate';
import { TRANSACTION_APPROVE_LOADING } from 'constants/crossChainTransfer';
import { EVM_TOKEN_ABI } from 'constants/wallet/EVM';
import { IWallet, WalletTypeEnum } from 'context/Wallet/types';
import { useAuthToken } from 'hooks/wallet/authToken';
import { getAelfMaxBalance } from 'pageComponents/CrossChainTransferPage/utils';
import { useCallback, useRef, useState } from 'react';
import { useLoading } from 'store/Provider/hooks';
import { TCreateTransferOrderResult } from 'types/api';
import {
  SendEVMTransactionParams,
  SendSolanaTransactionParams,
  SendTONTransactionParams,
  SendTRONTransactionParams,
} from 'types/wallet';
import { isDIDAddressSuffix, removeELFAddressSuffix } from 'utils/aelf/aelfBase';
import { isAuthTokenError } from 'utils/api/error';
import { createTransferOrder } from 'utils/api/transfer';
import { isAelfChain } from 'utils/wallet';

export function useCheckMaxBalance() {
  const [isBalanceNotEnoughTip, setIsBalanceNotEnoughTip] = useState(false);

  const checkMaxBalance = useCallback(
    async ({
      tokenSymbol,
      amount,
      fromBalance,
      fromNetwork,
      aelfTransactionFee,
      fromAccount,
    }: {
      tokenSymbol: string;
      amount?: string;
      fromBalance?: string;
      fromNetwork?: string;
      aelfTransactionFee?: string;
      fromAccount?: string | null;
    }) => {
      let _maxBalance = fromBalance;
      if (isAelfChain(fromNetwork || '') && tokenSymbol === 'ELF') {
        _maxBalance = await getAelfMaxBalance({
          balance: fromBalance || '',
          aelfFee: aelfTransactionFee,
          fromNetwork,
          tokenSymbol,
          account: fromAccount || '',
        });
      }
      const res = !!amount && (!_maxBalance || ZERO.plus(_maxBalance).lt(amount));
      setIsBalanceNotEnoughTip(res);
    },
    [],
  );

  return { isBalanceNotEnoughTip, checkMaxBalance };
}

export const DefaultWithdrawOrderResponse: TCreateTransferOrderResult = { orderId: '' };

export function useCreateOrderOtherNetwork(fromWallet?: IWallet, fromWalletType?: WalletTypeEnum) {
  const { setLoading } = useLoading();
  const { getAuthToken, queryAuthToken } = useAuthToken(fromWallet, fromWalletType);

  const authTokenRef = useRef('');
  const orderResultRef = useRef(DefaultWithdrawOrderResponse);

  const createOtherNetworkOrder = useCallback(
    async ({
      amount,
      fromNetwork,
      toNetwork,
      tokenSymbol,
      fromAddress,
      toAddress,
    }: {
      amount: string;
      fromNetwork: string;
      toNetwork: string;
      tokenSymbol: string;
      fromAddress: string;
      toAddress: string;
    }) => {
      // get etransfer jwt
      const authToken = await getAuthToken(true);
      authTokenRef.current = authToken;

      setLoading(true, { text: TRANSACTION_APPROVE_LOADING });

      // create order id
      const createTransferOrderParams = {
        amount: amount,
        fromNetwork: fromNetwork,
        toNetwork: toNetwork,
        fromSymbol: tokenSymbol,
        toSymbol: tokenSymbol,
        fromAddress: isDIDAddressSuffix(fromAddress)
          ? removeELFAddressSuffix(fromAddress)
          : fromAddress,
        toAddress: isDIDAddressSuffix(toAddress) ? removeELFAddressSuffix(toAddress) : toAddress,
      };

      try {
        orderResultRef.current = await createTransferOrder(createTransferOrderParams, authToken);
      } catch (error) {
        orderResultRef.current = DefaultWithdrawOrderResponse;
        if (isAuthTokenError(error)) {
          const _authToken = await queryAuthToken(true);
          authTokenRef.current = _authToken;
          orderResultRef.current = await createTransferOrder(createTransferOrderParams, _authToken);
        } else {
          throw error;
        }
      }
    },
    [getAuthToken, queryAuthToken, setLoading],
  );

  const sendOtherNetworkTransfer = useCallback(
    async ({
      amount,
      fromNetwork,
      fromWallet,
      decimalsFromWallet,
      tokenContractAddress,
    }: {
      amount: string;
      fromNetwork: string;
      fromWallet: IWallet;
      decimalsFromWallet?: string | number;
      tokenContractAddress?: string;
    }) => {
      let params: any;
      if (fromWallet.walletType === WalletTypeEnum.EVM) {
        params = {
          network: fromNetwork,
          tokenContractAddress: tokenContractAddress,
          toAddress: orderResultRef.current.address,
          tokenAbi: EVM_TOKEN_ABI,
          amount: amount,
          decimals: Number(decimalsFromWallet),
        } as SendEVMTransactionParams;
      } else if (fromWallet.walletType === WalletTypeEnum.SOL) {
        params = {
          tokenContractAddress: tokenContractAddress,
          toAddress: orderResultRef.current.address,
          amount: amount,
          decimals: decimalsFromWallet,
        } as SendSolanaTransactionParams;
      } else if (fromWallet.walletType === WalletTypeEnum.TON) {
        params = {
          tokenContractAddress: tokenContractAddress,
          toAddress: orderResultRef.current.address,
          amount: Number(amount),
          decimals: decimalsFromWallet,
          orderId: orderResultRef.current.orderId,
        } as SendTONTransactionParams;
      } else {
        // TRON
        params = {
          tokenContractAddress: tokenContractAddress,
          toAddress: orderResultRef.current.address,
          amount: Number(amount),
        } as SendTRONTransactionParams;
      }
      if (!fromWallet?.sendTransaction) return;

      const sendTransferResult = await fromWallet?.sendTransaction(params);

      return sendTransferResult;
    },
    [],
  );

  return { authTokenRef, orderResultRef, createOtherNetworkOrder, sendOtherNetworkTransfer };
}
