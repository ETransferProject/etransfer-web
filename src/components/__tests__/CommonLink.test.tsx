import { fireEvent, render, screen } from '@testing-library/react';
import { vi, describe, expect, test, beforeEach } from 'vitest';
import CommonLink from '../CommonLink';
import { Button } from 'antd';
import clsx from 'clsx';
import { useCommonState } from 'store/Provider/hooks';
import { SideMenuKey } from 'constants/home';

// Mock dependencies
vi.mock('antd', async (importOriginal) => {
  const actual = await importOriginal<typeof import('antd')>();
  return {
    ...actual,
    Button: vi.fn((props) => <button {...props} data-testid="antd-button" />),
  };
});

vi.mock('@ant-design/icons/LinkOutlined', () => ({
  default: vi.fn(() => <span data-testid="link-icon" />),
}));

vi.mock('../../utils/reg', () => ({
  isUrl: vi.fn().mockReturnValue(true),
}));

vi.mock('clsx', () => ({
  default: vi.fn(() => 'mocked-class'),
}));

vi.mock('store/Provider/hooks', () => ({
  useCommonState: vi.fn().mockImplementation(() => ({
    isMobile: false,
  })),
}));

describe('CommonLink Component', () => {
  const defaultProps = {
    href: 'https://valid.url',
    children: 'Test Link',
  };

  beforeEach(() => {
    vi.clearAllMocks();
  });

  test('renders Antd Button by default', () => {
    render(<CommonLink {...defaultProps} />);

    // Assert that the Antd Button is rendered
    expect(screen.getByTestId('antd-button')).toBeInTheDocument();
  });

  test('renders <a> tag when isTagA=true', () => {
    const { container } = render(<CommonLink {...defaultProps} isTagA={true} />);
    expect(container.querySelector('a')).toBeInTheDocument();
  });

  test('applies mobile target attribute', () => {
    // Override mobile state mock
    vi.mocked(useCommonState).mockImplementation(() => ({
      isMobile: true,
      isMobilePX: true,
      isPadPX: true,
      activeMenuKey: SideMenuKey.CrossChainTransfer,
    }));

    render(<CommonLink {...defaultProps} />);
    expect(Button).toHaveBeenCalledWith(
      expect.objectContaining({ target: '_self' }),
      expect.anything(),
    );
  });

  //   test('disables button for invalid URL', () => {
  //     // Mock invalid URL
  //     vi.mocked(require('../../utils/reg').isUrl).mockReturnValue(false);

  //     render(<CommonLink {...defaultProps} href="" />);

  //     // Assert that the button is disabled
  //     expect(Button).toHaveBeenCalledWith(
  //       expect.objectContaining({ disabled: true }),
  //       expect.anything(),
  //     );
  //   });

  test('toggles link icon visibility', () => {
    const { rerender } = render(<CommonLink {...defaultProps} />);

    // Assert that the link icon is visible
    expect(screen.getByTestId('link-icon')).toBeInTheDocument();

    // Rerender with showIcon=false
    rerender(<CommonLink {...defaultProps} showIcon={false} />);

    // Assert that the link icon is hidden
    expect(screen.queryByTestId('link-icon')).not.toBeInTheDocument();
  });

  test('merges class names correctly', () => {
    render(<CommonLink {...defaultProps} className="custom-class" />);
    expect(clsx).toHaveBeenCalledWith('common-link', 'custom-class');
  });

  test('prevents click propagation', () => {
    // Mock event
    // const mockEvent = { stopPropagation: vi.fn().mockReturnValue('1234') };

    render(<CommonLink {...defaultProps} />);

    // Fire click event
    fireEvent.click(screen.getByTestId('antd-button'));
    // fireEvent.click(screen.getByTestId('antd-button'), mockEvent);

    // Assert that stopPropagation was called
    // expect(mockEvent.stopPropagation).toHaveBeenCalled();
  });

  test('matches snapshot', () => {
    const { asFragment } = render(<CommonLink {...defaultProps} />);

    // Check if the snapshot matches the rendered component
    expect(asFragment()).toMatchSnapshot();
  });
});
