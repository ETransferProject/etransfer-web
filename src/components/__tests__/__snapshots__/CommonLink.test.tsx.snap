// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CommonLink Component > matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="mocked-class"
    data-testid="antd-button"
    href="https://valid.url"
    target="_self"
    type="link"
  >
    <span
      data-testid="link-icon"
    />
    Test Link
  </button>
</DocumentFragment>
`;
